project/src/main/webapp/gmail.js

```diff
@@ -12,193 +12,206 @@
-                        const messageRequest = gapi.client.gmail.users.messages.get({
+    /** Retrieve messages using hardcoded queries and the signed-in email. */
+    listMessages() {
+        this.emailObjects = {};
+        const listMessagePromise = new Promise((resolve) => {
+            const promiseArray = [];
+            let getPageOfMessages = (request, result) => {
+                request.execute(resp => {
+                    result = result.concat(resp.messages);
+                    var nextPageToken = resp.nextPageToken;
+                    if (nextPageToken) {
+                        request = gapi.client.gmail.users.messages.list({
                             'userId': 'me',
-                            'id': result[i].id,
-                            'format': "raw"
+                            'pageToken': nextPageToken,
+                            'q': this.question
                         });
-                        var messageID = result[i].id;
-                        messageRequest.execute(response => {
-                            const decodedEmail = atob(response.raw.replace(/-/g, '+').replace(/_/g, '/')); 
-                            const emailBodyStartIndex = decodedEmail.indexOf("Content-Type: text/plain; charset=\"UTF-8\"");
-                            const emailBodyEndIndex = (decodedEmail.substring(emailBodyStartIndex)).indexOf("Content-Type: text/html; charset=\"UTF-8\"")
-                            const emailBodyValue = decodedEmail.substring(emailBodyStartIndex, emailBodyStartIndex+emailBodyEndIndex);
-                            const emailDateStartIndex = decodedEmail.indexOf("Date:");
-                            const emailDateEndIndex = decodedEmail.substring(emailDateStartIndex).indexOf("Message-ID");
-                            var emailDateValue = decodedEmail.substring(emailDateStartIndex, emailDateStartIndex+emailDateEndIndex);
-                            if (emailDateValue.indexOf("X-Notifications") != -1) {
-                                emailDateValue = emailDateValue.substring(0, emailDateValue.indexOf("X-Notifications"));
-                            }
-                            var emailObject = {};
-                            emailObject[messageID] = {emailDate: emailDateValue, emailBody: emailBodyValue}; 
-                            if (emailBodyValue.indexOf("format=flowed") == -1 && emailDateValue.length > 1) {
-                                list[list.length] = emailObject;
+                        getPageOfMessages(request, result);
+                    } else {
+                        for (var i = 0; i < result.length; i++) {
+                            if (this.hasEmails(result)) {
+                                promiseArray.push(this.getMessage(result[i].id));
                             }
-                        });
-                    }
+                        }
+                        Promise.all(promiseArray).then(() => resolve()
```

Comment:
Remove this line - I believe this will cause listMessagePromise to resolve too early in some cases.

Reply:
If I remove this line the emailObjects never get sent to the emailParse. What cases would cause this to resolve too early?